extensions/amp-story-shopping/0.1/amp-story-shopping-config.js

```diff
@@ -1,19 +1,173 @@
  * @typedef {{
  *  items: !Array<!ShoppingConfigDataDef>,
  * }}
  */
 let ShoppingConfigResponseDef;
 
+/**
+ * Used for keeping track of intermediary invalid config results within Objects or Arrays of Objects.
+ * @private {boolean}
+ */
+let isValidConfigSection_ = true;
+
+/**
+ * Validates an Object using the validateConfig function.
+ * @param {?Object=} validation
+ * @return {boolean}
+ */
+function createValidateConfigObject(validation) {
+  return (field, value) => {
+    isValidConfigSection_ &&= validateConfig(value, validation, field + ' ');
+  };
+}
+
+/**
+ * Validates an Array of Objects using the validateConfig function.
+ * @param {?Object=} validation
+ * @return {boolean}
+ */
+function createValidateConfigArray(validation) {
+  return (field, value) => {
+    for (const item of value) {
+      isValidConfigSection_ &&= validateConfig(item, validation, field + ' ');
+    }
+  };
+}
+
+/**
+ * Validates if a required field exists for shopping config attributes
+ * @param {string} field
+ * @param {?string=} value
+ */
+export function validateRequired(field, value) {
+  if (value === undefined) {
+    throw Error(`Field ${field} is required.`);
+  }
+}
+
+/**
+ * Validates if string type for shopping config attributes
+ * @param {string} field
+ * @param {?string=} str
+ */
+export function validateString(field, str) {
+  if (typeof str !== 'string') {
+    throw Error(`${field} ${str} is not a string.`);
+  }
+}
+
+/**
+ * Validates number in shopping config attributes
+ * @param {string} field
+ * @param {?number=} number
+ */
+export function validateNumber(field, number) {
+  if (
+    (typeof number === 'string' && !/^[0-9.,]+$/.test(number)) ||
+    (typeof number !== 'string' && typeof number !== 'number')
+  ) {
+    throw Error(`Value ${number} for field ${field} is not a number`);
+  }
+}
+
+/**
+ * Validates url of shopping config attributes
+ * @param {string} field
+ * @param {?Array<string>=} url
+ */
+export function validateURLs(field, url) {
+  if (url === undefined) {
+    return;
+  }
+
+  const urls = Array.isArray(url) ? url : [url];
+
+  urls.forEach((url) => {
+    assertHttpsUrl(url, `amp-story-shopping-config ${field}`);
+  });
+}
+
+/**
+ * Validates the shopping config of a single product.
+ * @param {!ShoppingConfigDataDef} shoppingConfig
+ * @param {!Object<string, !Array<function>>} validationObject
+ * @param {?string} optParentFieldName
+ * @return {boolean}
+ */
+export function validateConfig(
+  shoppingConfig,
+  validationObject = productValidationConfig,
+  optParentFieldName = ''
+) {
+  let isValidConfig = true;
+
+  Object.keys(validationObject).forEach((configKey) => {
+    const validationFunctions = validationObject[configKey];
+    validationFunctions.forEach((fn) => {
+      try {
+        /* This check skips optional attribute validation */
+        if (
+          shoppingConfig[configKey] !== undefined ||
+          validationFunctions.includes(validateRequired)
+        ) {
```

Comment:
Does this mean that optional attributes go through no validation at all? If so, then should we just remove all optional attributes from `productValidationConfig` since they aren't being used?

Reply:
I should've been more clear with this comment, I'll update it:
`This check will skip optional attribute validation checks when they are not present`
Notice the `shoppingConfig[configKey] !== undefined`, this makes sure that the optional validation checks are run, assuming that they are present in the config.

Reply:
However, Optional fields that are invalid must not block rendering. 

`isValidConfig = false;` should not be set when an optional config field is present, and validateRequired is not in the array. I'm adding in some additional checks for this.